From 372ee3723015a8388fdbe4f183504202dade03f0 Mon Sep 17 00:00:00 2001
From: Thinkofdeath <thinkofdeath@spigotmc.org>
Date: Thu, 26 Jun 2014 22:48:09 +0100
Subject: [PATCH] Display unhandled damage sources


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index e6377c1..008b5e9 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -536,7 +536,7 @@ public class CraftEventFactory {
             return callEntityDamageEvent(null, entity, cause, modifiers);
         }
 
-        throw new RuntimeException("Unhandled entity damage");
+        throw new RuntimeException("Unhandled entity damage " + source.p()); // Spigot - getTranslationIndex
     }
 
     private static EntityDamageEvent callEntityDamageEvent(Entity damager, Entity damagee, DamageCause cause, Map<DamageModifier, Double> modifiers) {
-- 
1.9.1

