From 41a545a3d34211416d2babe064e6dab79189c8da Mon Sep 17 00:00:00 2001
From: geNAZt <fabian.fassbender42@googlemail.com>
Date: Tue, 8 Jul 2014 00:04:00 +0200
Subject: [PATCH] Fix player Skulls for 1.7.2 Clients


diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 9a5b235..2b59690 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -73,6 +73,7 @@ public class TileEntitySkull extends TileEntity {
 
             GameProfileSerializer.serialize(nbttagcompound1, this.j);
             nbttagcompound.set("Owner", nbttagcompound1);
+            nbttagcompound.setString("ExtraType", nbttagcompound1.getString("Name")); // Spigot
         }
     }
 
-- 
1.9.1

