From d716c43e9a202abe97e245ee42114fa855268394 Mon Sep 17 00:00:00 2001
From: geNAZt <fabian.fassbender42@googlemail.com>
Date: Tue, 8 Jul 2014 00:04:00 +0200
Subject: [PATCH] Fix player Skulls for 1.7.2 Clients


diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 255e0de..1441239 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -37,6 +37,7 @@ public class TileEntitySkull extends TileEntity {
 
             GameProfileSerializer.a(nbttagcompound1, this.j);
             nbttagcompound.set("Owner", nbttagcompound1);
+            nbttagcompound.setString("ExtraType", nbttagcompound1.getString("Name")); // Spigot
         }
     }
 
-- 
1.9.1

