From 47fe68bfe9552508211f663ee77c1a4efe0dd710 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sun, 10 Aug 2014 11:01:02 +1000
Subject: [PATCH] Force Close Inventory On Move


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 188393c..3f4268a 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -489,6 +489,12 @@ public class PlayerConnection implements PacketPlayInListener {
 
     public void a(PacketPlayInBlockDig packetplayinblockdig) {
         if (this.player.dead) return; // CraftBukkit
+        // Spigot start
+        if ( player.activeContainer != player.defaultContainer )
+        {
+            getPlayer().closeInventory();
+        }
+        // Spigot end
         WorldServer worldserver = this.minecraftServer.getWorldServer(this.player.dimension);
 
         this.player.v();
@@ -668,6 +674,12 @@ public class PlayerConnection implements PacketPlayInListener {
         }
 
         if (flag) {
+            // Spigot Start
+            if ( player.activeContainer != player.defaultContainer )
+            {
+                getPlayer().closeInventory();
+            }
+            // Spigot End
             this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(i, j, k, worldserver));
             if (l == 0) {
                 --j;
@@ -1118,6 +1130,12 @@ public class PlayerConnection implements PacketPlayInListener {
             return;
         }
         // Spigot End
+        // Spigot start
+        if ( player.activeContainer != player.defaultContainer )
+        {
+            getPlayer().closeInventory();
+        }
+        // Spigot end
 
         this.player.v();
         if (entity != null) {
-- 
1.9.1

